pl_bolts/datamodules/experience_source.py

```diff
@@ -29,174 +31,247 @@ def __iter__(self) -> Iterable:
-        self.state = self.env.reset()
-
-    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
-        """Takes a single step through the environment"""
-        action = self.agent(self.state, device)
-        new_state, reward, done, _ = self.env.step(action)
-        experience = Experience(
-            state=self.state,
-            action=action,
-            reward=reward,
-            new_state=new_state,
-            done=done,
-        )
-        self.state = new_state
-
-        if done:
-            self.state = self.env.reset()
-
-        return experience, reward, done
-
-    def run_episode(self, device: torch.device) -> float:
-        """Carries out a single episode and returns the total reward. This is used for testing"""
-        done = False
-        total_reward = 0
-
-        while not done:
-            _, reward, done = self.step(device)
-            total_reward += reward
 
-        return total_reward
+    def stepper(self) -> Experience:
+        """Iterable method that yields steps from the experience source"""
+        raise NotImplementedError("ExperienceSource has no stepper method implemented")
 
 
-class NStepExperienceSource(ExperienceSource):
-    """Expands upon the basic ExperienceSource by collecting experience across N steps"""
+class ExperienceSource(BaseExperienceSource):
+    """
+    Experience source class handling single and multiple environment steps
+
+    Args:
+        env: Environment that is being used
+        agent: Agent being used to make decisions
+        n_steps: Number of steps to return from each environment at once
+    """
 
-    def __init__(self, env, agent, n_steps: int = 1, gamma: float = 0.99):
+    def __init__(self, env, agent, n_steps: int = 1) -> None:
         super().__init__(env, agent)
-        self.gamma = gamma
+
+        if isinstance(env, (list, tuple)):
+            self.pool = env
+        else:
+            self.pool = [env]
+
         self.n_steps = n_steps
-        self.n_step_buffer = deque(maxlen=n_steps)
+        self.total_rewards = []
+        self.total_steps = []
+        self.states = []
+        self.histories = []
+        self.cur_rewards = []
+        self.cur_steps = []
+        self.iter_idx = 0
 
-    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
-        """
-        Takes an n-step in the environment
+        self.init_envs()
+
+    def stepper(self, device: torch.device) -> Tuple[Experience]:
+        """Experience Source iterator yielding Tuple of experiences for n_steps. These come from the pool
+        of environments provided by the user.
+
+        Args:
+            device: current device to be used for executing experience steps
 
         Returns:
-            Experience
+            Tuple of Experiences
         """
-        exp = self.n_step(device)
+        while True:
+
+            # get actions for all envs
+            actions = self.env_actions(device)
```

Comment:
I still haven't worked on this with DP/DDP in mind. But my reasoning is that the device is set based on where the stepper is called. So it could be run across multiple devices in parallel. However in that case, the environments need to be split up. More work needs to be done on this